R/entities_objects.R

```diff
@@ -0,0 +1,114 @@
+# Contains functions to parse the objects described here: 
+# https://developer.twitter.com/en/docs/twitter-api/v1/data-dictionary/object-model/entities
+
+# <https://developer.twitter.com/en/docs/twitter-api/v1/data-dictionary/object-model/entities#hashtags>
+hashtags <- function(x) {
+  if (NROW(x) == 0) {
+    return(data.frame(text = NA, indices = I(list(NA)), 
```

Comment:
Is it correct that a zero row input yields a one row output?

Reply:
Yes, the reason is to not omit anything for later easier parsing and linking between hasthags and tweets. There is also the problem when a user hasn't tweeted if the information about the tweets of that users and other users are collected it would break the tweet_data extraction see #574 for such a report.

Reply:
Right, I think you always want to return a data frame here, but are you sure it should have one row and not zero rows?

Reply:
Yes I'm sure. What is your concern? Filling the object with mostly empty data?

Reply:
Just generally reasoning about this function. I don't understand how all the pieces fit together but this feels like a code smell.

Reply:
I made some tests and it seems that mixing `NA`s and `data.frames` would be correctly handled by `rbind` unless there are nested `data.frame`s. I'll wait a bit to merge the PR but I can't think of any workaround that won't complicate more the workflow. 

<details>
<summary>Tests</summary>

``` r
``` r
l <- list(a = NA, b = data.frame(a = "b", b = "a"), c = data.frame(a = c("C", "D"), b = c("B", "C")))
do.call(rbind, l)
#>        a    b
#> a   <NA> <NA>
#> b      b    a
#> c.1    C    B
#> c.2    D    C
l <- list(a = data.frame(a = "b", b = "a"), b = NA, c = data.frame(a = c("C", "D"), b = c("B", "C")))
do.call(rbind, l)
#>        a    b
#> a      b    a
#> b   <NA> <NA>
#> c.1    C    B
#> c.2    D    C
l <- list(a = data.frame(a = "b", b = "a"), b = NA, c = data.frame(a = "C", b = "B"))
do.call(rbind, l)
#>      a    b
#> a    b    a
#> b <NA> <NA>
#> c    C    B
l <- list(a = data.frame(a = "b", b = "a"), b = data.frame(a = c("C", "D"), b = c("B", "C")), c = NA)
do.call(rbind, l)
#>        a    b
#> a      b    a
#> b.1    C    B
#> b.2    D    C
#> c   <NA> <NA>
library("tibble")
l <- list(a = NA, b = tibble(a = "b", b = "a"), c = tibble(a = c("C", "D"), b = c("B", "C")))
do.call(rbind, l)
#> # A tibble: 4 x 2
#>   a     b    
#> * <chr> <chr>
#> 1 <NA>  <NA> 
#> 2 b     a    
#> 3 C     B    
#> 4 D     C
l <- list(a = tibble(a = "b", b = "a"), b = NA, c = tibble(a = c("C", "D"), b = c("B", "C")))
do.call(rbind, l)
#> # A tibble: 4 x 2
#>   a     b    
#> * <chr> <chr>
#> 1 b     a    
#> 2 <NA>  <NA> 
#> 3 C     B    
#> 4 D     C
l <- list(a = tibble(a = "b", b = "a"), b = NA, c = tibble(a = "C", b = "B"))
do.call(rbind, l)
#> # A tibble: 3 x 2
#>   a     b    
#> * <chr> <chr>
#> 1 b     a    
#> 2 <NA>  <NA> 
#> 3 C     B
l <- list(a = tibble(a = "b", b = "a"), b = tibble(a = c("C", "D"), b = c("B", "C")), c = NA)
do.call(rbind, l)
#> # A tibble: 4 x 2
#>   a     b    
#> * <chr> <chr>
#> 1 b     a    
#> 2 C     B    
#> 3 D     C    
#> 4 <NA>  <NA>


l <- list(a = NA, b = data.frame(a = I(data.frame(c = 1)), b = "a"), c = data.frame(a = I(data.frame(c = c(2, 3))), b = c("B", "C")))
do.call(rbind, l)
#> Warning: non-unique value when setting 'row.names': '1'
#> Error in `.rowNamesDF<-`(x, value = value): duplicate 'row.names' are not allowed
l <- list(a = data.frame(a = I(data.frame(c = 1)), b = "a"), b = NA, c = data.frame(a = I(data.frame(c = c(2, 3))), b = c("B", "C")))
do.call(rbind, l)
#> Warning: non-unique values when setting 'row.names': '1', '2'
#> Error in `.rowNamesDF<-`(x, value = value): duplicate 'row.names' are not allowed
l <- list(a = data.frame(a = I(data.frame(c = 1)), b = "a"), b = NA, c = data.frame(a = I(data.frame(c = c(2))), b = "B"))
do.call(rbind, l)
#> Warning: non-unique value when setting 'row.names': '1'
#> Error in `.rowNamesDF<-`(x, value = value): duplicate 'row.names' are not allowed
l <- list(a = data.frame(a = I(data.frame(c = 1)), b = "a"), b = data.frame(a = I(data.frame(c = c(2, 3))), b = c("B", "C")), c = NA)
do.call(rbind, l)
#> Warning: non-unique value when setting 'row.names': '1'
#> Error in `.rowNamesDF<-`(x, value = value): duplicate 'row.names' are not allowed
l <- list(a = NA, b = tibble(a = I(data.frame(c = 1)), b = "a"), c = tibble(a = I(data.frame(c = c(2, 3))), b = c("B", "C")))
do.call(rbind, l)
#> Warning: non-unique value when setting 'row.names': '1'
#> Error in `.rowNamesDF<-`(x, value = value): duplicate 'row.names' are not allowed
l <- list(a = tibble(a = I(data.frame(c = 1)), b = "a"), b = NA, c = tibble(a = I(data.frame(c = c(2, 3))), b = c("B", "C")))
do.call(rbind, l)
#> Warning: non-unique values when setting 'row.names': '1', '2'
#> Error in `.rowNamesDF<-`(x, value = value): duplicate 'row.names' are not allowed
l <- list(a = tibble(a = I(data.frame(c = 1)), b = "a"), b = NA, c = tibble(a = I(data.frame(c = c(2))), b = "B"))
do.call(rbind, l)
#> Warning: non-unique value when setting 'row.names': '1'
#> Error in `.rowNamesDF<-`(x, value = value): duplicate 'row.names' are not allowed
l <- list(a = tibble(a = I(data.frame(c = 1)), b = "a"), b = tibble(a = I(data.frame(c = c(2, 3))), b = c("B", "C")), c = NA)
do.call(rbind, l)
#> Warning: non-unique value when setting 'row.names': '1'
#> Error in `.rowNamesDF<-`(x, value = value): duplicate 'row.names' are not allowed
```

<sup>Created on 2021-07-13 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

</details>

Reply:
I'm probably missing something, but I don't see you creating an zero row data frames in your tests?  But I don't really understand how all of this data flows together, and there's likely to be something upstream that means my suggestion wouldn't work.